adjust

```diff
@@ -0,0 +1,163 @@
+                    assert len(bad_keys) < 1, \
+                        "comp {}: range setting {} config was malformed, contained unknown key(s) {}".format(c_name, s_name, ', '.join(bad_keys))
+
+                    assert isinstance(s_data.get('min'), (int, float)), \
+                        'comp {}: range setting {} config was malformed; min value is required, must be a number. Found {}.'.format(c_name, s_name, s_data.get('min'))
+                    assert isinstance(s_data.get('max'), (int, float)), \
+                        'comp {}: range setting {} config was malformed; max value is required, must be a number. Found {}.'.format(c_name, s_name, s_data.get('max'))
+                    assert isinstance(s_data.get('step'), (int, float)), \
+                        'comp {}: range setting {} config was malformed; step value is required, must be a number. Found {}.'.format(c_name, s_name, s_data.get('step'))
+                    # Optional param, pass if missing
+                    assert isinstance(s_data.get('default', 0), (int, float)), \
+                        'comp {}: range setting {} config was malformed; default must be a number when provided. Found {}.'.format(c_name, s_name, s_data.get('default'))
+                        
+                    assert s_data['min'] <= s_data['max'], \
+                        'comp {}: range setting {} config was malformed; supplied min is higher than max'.format(c_name, s_name)
+
+                    if s_data['min'] != s_data['max']:
+                        assert s_data['step'] != 0, \
+                            'comp {}: range setting {} config was malformed; step cannot be zero when min != max.'.format(c_name, s_name)
+                        assert s_data['step'] > 0, \
+                            'comp {}: range setting {} config was malformed; step must be a positive number.'.format(c_name, s_name)
+
+                        c = (s_data['max'] - s_data['min']) / float(s_data['step'])
+                        if not math.isclose(c, round(c, 0), abs_tol = 1/1024):
+                            raise Exception(
+                                'comp {}: range setting {} config was malformed; step value must allow to get from {} to {} in equal steps of {}.'.format(
+                                    c_name, s_name, s_data['min'], s_data['max'], s_data['step']))
+
+        return cfg
+
+
+    def query(self):
+
+        cfg = self._parse_cfg()
+
+        q = {
+            "components": cfg["application"]["components"]
+        }
+        if "annotations" in cfg["application"]:
+            q["annotations"] = cfg["application"]["annotations"]
+
+        # Get state, merge
+        state = state_store.get_state()
+
+        if state:
+            assert ("application" in state), "Invalid state: missing application key"
+
+        for c_name, c_data in cfg["application"]["components"].items():
```

Comment:
"application" may include other things besides "components". A better approach would be to take the "application" key from config, replace anything that is not "components" with a matching key from the statestore (if present), then update components with the value from statestore (if present).